dask/dataframe/io/tests/test_io.py

```diff
@@ -354,36 +355,68 @@ def test_from_pandas_with_datetime_index():
 
 def test_DataFrame_from_dask_array():
     x = da.ones((10, 3), chunks=(4, 2))
-
+    pdf = pd.DataFrame(np.ones((10, 3)), columns=["a", "b", "c"])
     df = dd.from_dask_array(x, ["a", "b", "c"])
-    assert isinstance(df, dd.DataFrame)
-    tm.assert_index_equal(df.columns, pd.Index(["a", "b", "c"]))
-    assert list(df.divisions) == [0, 4, 8, 9]
-    assert (df.compute(scheduler="sync").values == x.compute(scheduler="sync")).all()
+    assert not hlg_layer_topological(df.dask, -1).is_materialized()
+    assert_eq(df, pdf)
 
     # dd.from_array should re-route to from_dask_array
     df2 = dd.from_array(x, columns=["a", "b", "c"])
-    assert isinstance(df, dd.DataFrame)
-    tm.assert_index_equal(df2.columns, df.columns)
-    assert df2.divisions == df.divisions
+    assert not hlg_layer_topological(df2.dask, -1).is_materialized()
+    assert_eq(df, df2)
+
+
+def test_DataFrame_from_dask_array_with_blockwise_ops():
+    x = da.ones((10, 3), chunks=(4, 2))
+    x *= 2
+    pdf = pd.DataFrame(np.ones((10, 3)) * 2, columns=["a", "b", "c"])
+    df = dd.from_dask_array(x, ["a", "b", "c"])
+    # None of the layers in this graph should be materialized, everything should
+    # be a HighLevelGraph still.
+    assert all(
+        not hlg_layer_topological(df.dask, i).is_materialized()
+        for i in range(len(df.dask.layers))
+    )
+    assert_eq(df, pdf)
+
+
+def test_create_index_is_blockwise():
+    x = da.ones((10, 3), chunks=(4, 2))
+    # The index will be a range from 0-10 with three partitions. Each partition index's
+    # starting value will be 4 higher than the previous one since chunks=(4, 2).
+    index_mapping = {
+        (0,): (0, 4),
+        (1,): (4, 8),
+        (2,): (8, 10),
+    }
+    df = dd.from_dask_array(x, ["a", "b", "c"])
+    layer = hlg_layer(df.dask, "from-dask-array")
+    # The layer to create the DataFrame should create the index using a BlockwiseDep
+    # If that is the case, then the io_deps dictionary will not be empty.
+    assert layer.io_deps
+    # Also test that the BlockwiseDepDict mapping attribute corresponds to the
+    # RangeIndex that will be created for `x`. There's only one item in the io_deps
+    # dictionary here, so we avoid a loop with this line. This is specific to
+    # BlockwiseDepDict, so it'll break if from_dask_array is refactored to use a
+    # different means of generating the index.
+    blk_dep = next(iter(layer.io_deps.values()))
+    assert blk_dep.mapping == index_mapping
```

Comment:
@rjzamora Thanks for the suggestions! I adopted them as-is, other than linting 😄 

I also added this test, which I'm about 80% sure is valuable. I wanted to assert that the layer creating the index is a HLG/blockwise, since that's what your change enabled. I think this test does that, but it does depend on the specific implementation using `BlockwiseDepDict` which is... not great? At least, it'll break if the implementation is changed, which IMO means it's not a great test. Happy for your thoughts though!

Reply:
I think I agree with your intuition that checking for `BlockwiseDepDict` usage is not ideal, because we may decide to use a different kind of `BlockwiseDep` or tweak the design in the future. I do like how thorough this test is, but my preference would be to simply assert that `isisntance(layer, Blockwise)` here, and to rely on the `assert_eq` in other tests to ensure that the index is being set correctly.

One other thing you can test is whether `dask.blockwise.optimize_blockwise` fuses everything down to one `Blockwise` layer.  For example:

```python
from dask.blockwise import Blockwise, optimize_blockwise

graph = optimize_blockwise(df.dask, df.__dask_keys__())
assert len(graph.layers) == 1
assert isinstance(hlg_layer(graph, "from-dask-array"), Blockwise)
```
(note: I didn't check that this works yet)

Reply:
I agree with both you and @rjzamora that delving into the specific shape of `BlockwiseDepDict` is not ideal, and is best left as an implementation detail.

`assert_eq` to verify the index still seems like a good idea here: if the range index construction is wrong then this would catch it.

> Unfortunately, the .is_materialized() is true even if the index creation is done non-Blockwise

I'm not sure I understand what you mean here, can you elaborate? We should be able to verify some things about the HLG before materialization.

Reply:
>I think it's not super important to really assert that the index is generated by a BlockwiseDep, but it also felt weird to apply a bunch of changes to the code and not check that they were doing what they were supposed to be doing.

Right - I would argue that it is not important at all that the index is generated by a BlockwiseDep, and so there is no reason to test this. We only care that from_dask_array produces a single Blockwise layer. The fact that we use BlockwiseDep for the index is just to make this possible.

Reply:
> I would argue that it is not important at all that the index is generated by a BlockwiseDep, and so there is no reason to test this.

Thanks @rjzamora, I think it makes sense to drop the test then.

> > Unfortunately, the .is_materialized() is true even if the index creation is done non-Blockwise
> 
> I'm not sure I understand what you mean here, can you elaborate? We should be able to verify some things about the HLG before materialization.

@ian-r-rose Sorry, that was a brain fart. I meant `hlg_layer_topological(df.dask, -1).is_materialized()` is _`False`_, so that `assert not ...` is `True` and the test passes. This happened even when I was constructing the low-level graph directly (`dsk = {} ...`), because the `from-dask-array` task hadn't been materialized even though its dependencies were created as a materialized graph. I don't recall what happens if I try to check _all_ the layers though, but I think the point is moot now since I agree with Rick that it makes sense to drop this test as being too specific to the implementation.